src/com/amazon/corretto/crypto/provider/EcParameters.java

```diff
@@ -0,0 +1,115 @@
+        if (paramSpec == null) {
+            throw new InvalidParameterSpecException("paramSpec must not be null");
+        }
+
+        String name = null;
+        if (paramSpec instanceof ECParameterSpec) {
+            name = EcUtils.getNameBySpec((ECParameterSpec) paramSpec);
+            // Earlier Java TLS implementations cache curve params by OID instead of human-readable "shortname".
+            // Specifying the shortname where OID is expected results in TLS handshake failures due to the server
+            // not being able to "find" the intended curve, tricking the server into believing it doesn't support any
+            // of the curves stipulated by the client.
+            //
+            // https://github.com/corretto/corretto-8/blob/235873fd43e5b7aa556d011f436e65a99c10c20a/jdk/src/share/classes/sun/security/ssl/SupportedGroupsExtension.java#L585-L607
+            // https://github.com/openjdk/jdk10u/blob/ef9178d7d8a4489640a31a1d0c88958724af5304/src/java.base/share/classes/sun/security/ssl/SupportedGroupsExtension.java#L188-L210
+            if (Utils.getJavaVersion() < 11) {
+                name = EcUtils.getOidFromName(name);
+            }
+        } else if (paramSpec instanceof ECGenParameterSpec) {
+            name = ((ECGenParameterSpec) paramSpec).getName();
+        } else if ("sun.security.util.ECKeySizeParameterSpec".equals(paramSpec.getClass().getName())) {
+            // OpenJDK's JCE sometimes passes ECKeySizeParameterSpec when initializing EC AlgorithmParameters. We can
+            // reference this class directly using instanceof up until Java 8, but in Java 9 it was moved to the
+            // java.base module and only exported to other JDK-internal modules. To work around this, we use reflection
+            // to get the key size and get the corresponding curve name from our own database. Future versions of java
+            // may restrict reflective access to private modules, so this functionality may break.
+            //
+            // https://github.com/corretto/corretto-11/blob/14c02261590b4dc01284888a7a51d39ff581ac8d/src/java.base/share/classes/sun/security/util/ECParameters.java#L121
+            // https://github.com/corretto/corretto-11/blob/14c02261590b4dc01284888a7a51d39ff581ac8d/src/java.base/share/classes/module-info.java#L301-L314
+            try {
+                Method getKeySize = paramSpec.getClass().getMethod("getKeySize");
+                Integer keySize = Integer.class.cast(getKeySize.invoke(paramSpec));
+                name = EcUtils.getNameByKeySize(keySize);
+            } catch (ReflectiveOperationException e) {
+                // pass, perhaps due to reflective access control restrictions. rely null check below to throw.
+            }
+        }
+
+        if (name == null) {
+            throw new InvalidParameterSpecException("Only ECParameterSpec and ECGenParameterSpec supported");
```

Comment:
Error message is potentially misleading. We can also get a null name by the EcUtils.get*() queries failing. Parameterizing the error message would be ideal, but I can see how that's tricky due to scoping here.

Reply:
Because ECKeySizeParameterSpec is a JDK-internal class, I consider the reflective check we do to be an implementation detail and thus intentionally don't want to expose it to callers. the only other two classes we handle are ECParameterSpec and ECGenParameterSpec (as articulated by this message). If any other class is specified, `name` will be null at the end of this conditional block and the error message will be thrown.